src/libraries/System.Private.CoreLib/src/System/Int128.cs

```diff
@@ -289,6 +293,10 @@ public static bool TryParse(ReadOnlySpan<char> s, NumberStyles style, IFormatPro
             if (~value._upper == 0)
             {
                 long lower = (long)value._lower;
+                if (lower >= 0)
+                {
+                    ThrowHelper.ThrowOverflowException();
+                }
```

Comment:
I'm not sure this is the best fix here.

What the check is meant to be doing is effectively determining "am I a 64-bit value so that I can rely on the regular checked conversion".

The issue here is that if the value is a large negative (upper is `AllBitsSet`) but `lower` then doesn't have its most significant bit set, we may see just `lower` as a small positive.

The entire logic could be replaced with this instead:
```csharp
if (value._upper != ((long)lower >> 63))
{
    ThrowHelper.ThrowOverflowException();
}
return checked((int)value._lower);
```

This works since all 63-bit positive values are in the range `0x0000_0000_0000_0000, 0x0000_0000_0000_0000` (0) to `0x0000_0000_0000_0000, 0x7FFF_FFFF_FFFF_FFFF` (long.MaxValue)

While all 63-bit negative values are in the range `0xFFFF_FFFF_FFFF_FFFF, 0xFFFF_FFFF_FFFF_FFFF` (-1) to `0xFFFF_FFFF_FFFF_FFFF, 0x8000_0000_0000_0000` (long.MinValue)

The issue was then simply that the logic wasn't checking the most significant bit of `lower` matched the entire bit pattern from `upper`

Reply:
Followed your suggestion but changed the last line.